iot-hub/Samples/device/MessageReceiveSample/MessageReceiveSample.cs

```diff
@@ -0,0 +1,108 @@
+            await ReceiveC2dMessagesPollingAndComplete(s_receiveTimeout);
+
+            // Now unsubscribe from the callback.
+            await _deviceClient.SetReceiveMessageHandlerAsync(null, _deviceClient);
+            Console.WriteLine($"\n{DateTime.Now}> Unsubscribed from receiving C2D messages over callback.");
+
+            // For Mqtt - since we have explicitly unsubscribed, we will need to resubscribe again
+            // before the device can begin receiving C2D messages.
+            if (_transportType == TransportType.Mqtt_Tcp_Only
+                || _transportType == TransportType.Mqtt_WebSocket_Only)
+            {
+                Message leftoverMessage = await _deviceClient.ReceiveAsync(s_sleepDuration).ConfigureAwait(false);
+            }
+
+            // Now wait to receive message from service.
+            // The message should be received on the polling ReceiveAsync() call and not on the callback.
+            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub for {s_receiveTimeout}...");
+            Console.WriteLine($"{DateTime.Now}> Use the Azure Portal IoT Hub blade or Azure IoT Explorer to send a message to this device.");
+            await ReceiveC2dMessagesPollingAndComplete(s_receiveTimeout);
+        }
+
+        private async Task ReceiveC2dMessagesPollingAndComplete(TimeSpan timeout)
+        {
+            var sw = new Stopwatch();
+            sw.Start();
+
+            Console.WriteLine($"{DateTime.Now}> Receiving C2D messages on the polling ReceiveAsync().");
+            while (sw.Elapsed < timeout)
+            {
+                using Message receivedMessage = await _deviceClient.ReceiveAsync(timeout);
+
+                if (receivedMessage == null)
+                {
+                    Console.WriteLine($"{DateTime.Now}> Polling ReceiveAsync() - no message received.");
+                    await Task.Delay(s_sleepDuration);
+                    continue;
+                }
+
+                Console.WriteLine($"{DateTime.Now}> Polling ReceiveAsync() - received message with Id={receivedMessage?.MessageId}");
+                ProcessReceivedMessage(receivedMessage);
+
+                await _deviceClient.CompleteAsync(receivedMessage);
+                Console.WriteLine($"{DateTime.Now}> Completed C2D message with Id={receivedMessage?.MessageId}.");
+            }
+
+            sw.Stop();
+        }
+
+        private async Task OnC2dMessageReceived(Message receivedMessage, object _)
+        {
+            Console.WriteLine($"{DateTime.Now}> C2D message callback - message received with Id={receivedMessage?.MessageId}.");
+            ProcessReceivedMessage(receivedMessage);
+
+            await _deviceClient.CompleteAsync(receivedMessage);
+            Console.WriteLine($"{DateTime.Now}> Completed C2D message with Id={receivedMessage?.MessageId}.");
```

Comment:
Why might received message be null?

Reply:
I don't see this was resolved.

Reply:
you are correct, the callback will never be invoked with a null message